### ISSUE TITLE
preCreateUser hook emitted at different location

### ISSUE THREAD

**Issue by GitHubUser4234** [nextcloud/server#5457]
Hi,

with commit [9212089151e827b648033a4812a5dcac5058b715 ](https://github.com/nextcloud/server/commit/9212089151e827b648033a4812a5dcac5058b715#diff-2ad387ea4134715ffda3bb0f12140f91) the location of the preCreateUser hook emission has changed. Unfortunately this causes some side-effects for projects that like to use a different backend when creating a user through Nextcloud's "Users" page. Before, one could write an app with a custom backend, and during user creation through Nextcloud's "Users" page the app would do the following: 

- remove the database backend in the preCreateUser hook 
- create a user by the custom backend
- re-add the database backend

As discussed with @nickvergessen in IRC, this feature can be brought back, thanks a lot!

**Comment by nickvergessen** [comment]
The dirty fix would be to bring this back:
```diff
diff --git a/lib/private/User/Manager.php b/lib/private/User/Manager.php
index 0477f23e55..a8de67de1e 100644
--- a/lib/private/User/Manager.php
+++ b/lib/private/User/Manager.php
@@ -284,9 +284,10 @@ class Manager extends PublicEmitter implements IUserManager {
         * @return bool|IUser the created user or false
         */
        public function createUser($uid, $password) {
+               $this->emit('\OC\User', 'preCreateUser', [$uid, $password]);
                foreach ($this->backends as $backend) {
                        if ($backend->implementsActions(Backend::CREATE_USER)) {
-                               return $this->createUserFromBackend($uid, $password, $backend);
+                               return $this->createUserFromBackend($uid, $password, $backend, false);
                        }
                }
 
@@ -297,10 +298,11 @@ class Manager extends PublicEmitter implements IUserManager {
         * @param string $uid
         * @param string $password
         * @param UserInterface $backend
+        * @param bool $emitPreHook
         * @return IUser|null
         * @throws \InvalidArgumentException
         */
-       public function createUserFromBackend($uid, $password, UserInterface $backend) {
+       public function createUserFromBackend($uid, $password, UserInterface $backend, $emitPreHook = true) {
                $l = \OC::$server->getL10N('lib');
 
                // Check the name for bad characters
@@ -331,7 +333,9 @@ class Manager extends PublicEmitter implements IUserManager {
                        throw new \InvalidArgumentException($l->t('The username is already being used'));
                }
 
-               $this->emit('\OC\User', 'preCreateUser', [$uid, $password]);
+               if ($emitPreHook) {
+                       $this->emit('\OC\User', 'preCreateUser', [$uid, $password]);
+               }
                $backend->createUser($uid, $password);
                $user = $this->getUserObject($uid, $backend);
                if ($user instanceof IUser) {
```

the question is whether we want to add a config for this instead. Or do something like "always check database backend last".
I would prefer both over this, since it is a cleaner way isntead of abusing some random facts which can break from time to time (like it happened)

**Comment by GitHubUser4234** [comment]
Thanks a lot @nickvergessen, I'll use the quick fix above to patch our NC12.

> the question is whether we want to add a config for the prefered user backend instead. Or do something like "always check database backend last".
> I would prefer both over this, since it is a cleaner way isntead of abusing some random facts which can break from time to time (like it happened)

Yeah, would be great if any of the suggested options would be available. While the first option sounds like the most appealing, the latter one could probably be applied with less effort? Anyhow, at the bottom line,  it would be desireable not to have to manually patch the core anymore 😹

**Comment by blizzz** [comment]
Since the behaviour changes, this can qualify as regression. Especially since any other way is technically a feature.

Nevertheless I agree having a proper way to prioritize user backends. A general rule to always check the local one last seems reasonable. After all, if you enable an extra one, it is typically the main source. And this we can  do without much other magic or even extra configuration options. Bringing this to back 12 instead of relocation the preHook as a fix should solve your issue @GitHubUser4234, right?

**Comment by GitHubUser4234** [comment]
> Nevertheless I agree having a proper way to prioritize user backends. A general rule to always check the local one last seems reasonable. After all, if you enable an extra one, it is typically the main source. And this we can do without much other magic or even extra configuration options. Bringing this to back 12 instead of relocation the preHook as a fix should solve your issue @GitHubUser4234, right?

Wow, this really sounds like a great option! No more need for removing/re-adding the database backend would make the whole approach look a lot less hacky. Sure, it has to be tested, but basically this solution would be very appreciated, thanks a lot @blizzz  :smile:

**Comment by MorrisJobke** [comment]
Fixed in 12.0.1 and 13.
